### ISSUE TITLE
Unescaped names are emitted during codegen

### ISSUE THREAD

**Issue by blast-hardcheese** [scalameta/scalameta#1661]
I'm generating code from arbitrary user input. Sometimes I get names that are not valid Scala syntax; it would be nice if scalameta correctly handled this case.
```
scala> import scala.meta._
import scala.meta._

scala> q"val `foo-bar`: Long = 5L"
res0: meta.Defn.Val = val foo-bar: Long = 5L

scala> res0.structure
res1: String = Defn.Val(Nil, List(Pat.Var(Term.Name("foo-bar"))), Some(Type.Name("Long")), Lit.Long(5L))

scala> q"val ${Pat.Var(Term.Name("`foo-bar`"))}: Long = 5L"
res2: meta.Defn.Val = val `foo-bar`: Long = 5L
```

Previously I've sort of hacked around this using [`.transform`](https://github.com/twilio/guardrail/blob/cbf9acd9e8ff226cc0f4bbf2f278669071126d5e/modules/codegen/src/main/scala/com/twilio/guardrail/SwaggerUtil.scala#L259-L267) and [some regex/string matching](https://github.com/twilio/guardrail/blob/cbf9acd9e8ff226cc0f4bbf2f278669071126d5e/modules/codegen/src/main/scala/com/twilio/guardrail/SwaggerUtil.scala#L269-L333), but this was intended to be a short/medium term workaround and now I'm running into stack depth issues.

This is using scalameta 3.7.4 and Scala 2.11/2.12

**Comment by blast-hardcheese** [comment]
Oh, in digging in here, it seems as though all my problems are from https://github.com/scalameta/scalameta/blob/4174e5a81d42ceb89758fd98dfdf2eb18a26dc6b/scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala#L123 not including all symbols as terms that needs to be escaped

**Comment by olafurpg** [comment]
Thanks for reporting! We are working on a new tree printer in https://github.com/scalacenter/scala-syntax that already improves on the current printer in several ways including

- more correctness in tricky corner cases for infix operators, symbolic identifiers and operator precedence/associativity
- line wrapping to maximum column width
- WIP support to preserve comments (even after `.transform`)

It's not yet published because comments are still incomplete. Unfortunately we haven't had resources to invest more in the project so it's been stalled since March. If you are on sbt 0.13 you can depend on it via source dependencies as is done here https://github.com/olafurpg/codegen-example/blob/756d8f0074fd3e07dcd8b0a1e347471cec1966f5/build.sbt#L6
Here is an example usage of the API https://github.com/olafurpg/codegen-example/blob/756d8f0074fd3e07dcd8b0a1e347471cec1966f5/codegen/src/main/scala/codegen/Main.scala

If you can give it a try and see if it's better than the current printer then I can investigate if we can publish the printer to Maven. I already know there's interest in using the printer in https://github.com/felixmulder/scalambda

**Comment by blast-hardcheese** [comment]
@olafurpg I am, in fact, running 0.13! Testing it out now. If it works, offering it as an alternative pretty printer would be fantastic, I've been deferring on a ticket for months around no comments in generated code 😄

**Comment by olafurpg** [comment]
Comments in generated code are a tough challenge but I think we have a cracked the nut in scala-syntax, it can both automatically preserve comments in existing code

https://github.com/scalacenter/scala-syntax/blob/7352163153066bf9594d63c8e78d62aaa9123656/tests/unit/src/test/scala/org/scalafmt/tests/CommentsSuite.scala

and also supports an API to manually attach comments to synthetic tree nodes

https://github.com/scalacenter/scala-syntax/blob/7352163153066bf9594d63c8e78d62aaa9123656/tests/unit/src/test/scala/org/scalafmt/tests/DocstringSuite.scala

Somebody just needs to find the time and motivation to grind through our `PreserveCommentsTest` suite and make sure that no comments in the original source get lost https://github.com/scalacenter/scala-syntax/blob/7352163153066bf9594d63c8e78d62aaa9123656/tests/slow/src/test/scala/org/scalafmt/internal/PreserveCommentsTest.scala 😉

**Comment by blast-hardcheese** [comment]
@olafurpg In attempting to test out https://github.com/scalacenter/scala-syntax , I'm having some issues -- as soon as I add the `dependsOn` call, my test runner immediately stops finding my tests. Do you have any insight here?

```
diff --git a/build.sbt b/build.sbt
index c1a3dd2..273915d 100644
--- a/build.sbt
+++ b/build.sbt
@@ -93,6 +93,10 @@ val codegenSettings = Seq(
   , offline := true
 )
 
+lazy val syntax = ProjectRef(uri(
+    "git://github.com/olafurpg/scala-syntax.git#9dc1da7e19991bfc95e6a874de48eb5b95e4a3a5"
+  ), "format")
+
 lazy val root = (project in file("."))
   .settings(
     libraryDependencies ++= testDependencies
@@ -104,6 +108,7 @@ lazy val codegen = (project in file("guardrail/modules/codegen"))
     (name := s"${projectName}-core") +:
     codegenSettings
   )
+  .dependsOn(syntax % "test->test;compile->compile")
 
 lazy val twilio = (project in file("modules/twilio"))
   .settings(
```

I tried using `RootProject` instead of `ProjectRef` in this way, but it seems as though this invocation is similar in effect without giving the impression that I have two project roots.

**Comment by blast-hardcheese** [comment]
Ah, figured it out -- that repo includes https://github.com/scalacenter/scala-syntax/blob/4b9593cfa3d79c2ea5f109db4c85557d0c5d9c10/project/FormatBuild.scala#L15-L17

**Comment by blast-hardcheese** [comment]
@olafurpg I just redefined
```
testFrameworks in ThisBuild := List(TestFramework("org.scalatest.tools.Framework", "org.scalatest.tools.ScalaTestFramework"), TestFramework("org.scalafmt.tests.CustomFramework"))
```
and my tests are passing now! Spot-checking after ripping all my `.transform` stuff shows everything is being escaped as expected now, which is really great.

**Comment by blast-hardcheese** [comment]
Whoops,
```
[error] (run-main-8) java.util.NoSuchElementException: None.get
java.util.NoSuchElementException: None.get
        at scala.None$.get(Option.scala:349)
        at scala.None$.get(Option.scala:347)
        at org.scalafmt.internal.tokens.SyntaxTokensUtils$.$anonfun$commaSeparated0$1(SyntaxTokensUtils.scala:102)
        at scala.collection.Iterator$$anon$10.next(Iterator.scala:457)
        at scala.collection.Iterator.foreach(Iterator.scala:944)
        at scala.collection.Iterator.foreach$(Iterator.scala:944)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
        at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:59)
        at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:50)
        at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:186)
        at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:44)
        at scala.collection.TraversableOnce.to(TraversableOnce.scala:310)
        at scala.collection.TraversableOnce.to$(TraversableOnce.scala:308)
        at scala.collection.AbstractIterator.to(Iterator.scala:1432)
        at scala.collection.TraversableOnce.toList(TraversableOnce.scala:294)
        at scala.collection.TraversableOnce.toList$(TraversableOnce.scala:294)
        at scala.collection.AbstractIterator.toList(Iterator.scala:1432)
        at org.scalafmt.internal.tokens.SyntaxTokensUtils$.commaSeparated0(SyntaxTokensUtils.scala:101)
        at org.scalafmt.internal.tokens.SyntaxTokensUtils$.commaSeparated(SyntaxTokensUtils.scala:90)
        at org.scalafmt.internal.tokens.SyntaxTokensTerm$XtensionTermTupleSyntax$.tokensComma$extension(SyntaxTokensTerm.scala:224)
        at org.scalafmt.internal.tokens.SyntaxTokensTerm$XtensionTermTupleSyntax$.$u002C$extension(SyntaxTokensTerm.scala:235)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:233)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:230)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dApply$1(TreeDocOps.scala:67)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalafmt.internal.TreeDocOps.dApply(TreeDocOps.scala:67)
        at org.scalafmt.internal.TreeDocOps.dApply$(TreeDocOps.scala:66)
        at org.scalafmt.internal.TreePrinter.dApply(TreePrinter.scala:58)
        at org.scalafmt.internal.TreeDocOps.dApplyParen(TreeDocOps.scala:60)
        at org.scalafmt.internal.TreeDocOps.dApplyParen$(TreeDocOps.scala:59)
        at org.scalafmt.internal.TreePrinter.dApplyParen(TreePrinter.scala:58)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:290)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:186)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:282)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:287)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:58)
        at org.scalafmt.internal.TreeDocOps.dBlockI(TreeDocOps.scala:103)
        at org.scalafmt.internal.TreeDocOps.dBlockI$(TreeDocOps.scala:88)
        at org.scalafmt.internal.TreePrinter.dBlockI(TreePrinter.scala:58)
        at org.scalafmt.internal.TreeDocOps.dBlock(TreeDocOps.scala:86)
        at org.scalafmt.internal.TreeDocOps.dBlock$(TreeDocOps.scala:85)
        at org.scalafmt.internal.TreePrinter.dBlock(TreePrinter.scala:58)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:225)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dBodyO$2(TreeDocOps.scala:163)
        at scala.Option.fold(Option.scala:158)
        at org.scalafmt.internal.TreeDocOps.dBodyO(TreeDocOps.scala:161)
        at org.scalafmt.internal.TreeDocOps.dBodyO$(TreeDocOps.scala:160)
        at org.scalafmt.internal.TreePrinter.dBodyO(TreePrinter.scala:58)
        at org.scalafmt.internal.TreeDocOps.dBody(TreeDocOps.scala:158)
        at org.scalafmt.internal.TreeDocOps.dBody$(TreeDocOps.scala:157)
        at org.scalafmt.internal.TreePrinter.dBody(TreePrinter.scala:58)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:497)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:58)
        at org.scalafmt.internal.TreeDocOps.dBlockI(TreeDocOps.scala:103)
        at org.scalafmt.internal.TreeDocOps.dBlockI$(TreeDocOps.scala:88)
        at org.scalafmt.internal.TreePrinter.dBlockI(TreePrinter.scala:58)
        at org.scalafmt.internal.TreeDocOps.dBlock(TreeDocOps.scala:86)
        at org.scalafmt.internal.TreeDocOps.dBlock$(TreeDocOps.scala:85)
        at org.scalafmt.internal.TreePrinter.dBlock(TreePrinter.scala:58)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:448)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:527)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:287)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:58)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:378)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:58)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:532)
        at org.scalafmt.internal.TreePrinter$.print(TreePrinter.scala:33)
        at com.twilio.guardrail.WriteTree$.$anonfun$unsafeWriteTree$1(WriteTree.scala:16)
```

**Comment by blast-hardcheese** [comment]
that exception was thrown on
```
Source(List(Pkg(Term.Select(Term.Select(Term.Name("tests"), Term.Name("dtos")), Term.Name("definitions")), List(Import(List(Importer(Term.Select(Term.Name("io"), Term.Name("circe")), List(Importee.Wildcard())))), Import(List(Importer(Term.Select(Term.Select(Term.Name("io"), Term.Name("circe")), Term.Name("syntax")), List(Importee.Wildcard())))), Import(List(Importer(Term.Select(Term.Select(Term.Select(Term.Name("io"), Term.Name("circe")), Term.Name("generic")), Term.Name("semiauto")), List(Importee.Wildcard())))), Import(List(Importer(Term.Select(Term.Name("cats"), Term.Name("implicits")), List(Importee.Wildcard())))), Import(List(Importer(Term.Select(Term.Select(Term.Name("com"), Term.Name("twilio")), Term.Name("sids")), List(Importee.Wildcard())))), Import(List(Importer(Term.Select(Term.Select(Term.Select(Term.Select(Term.Name("com"), Term.Name("twilio")), Term.Name("service")), Term.Name("encoding")), Term.Name("json")), List(Importee.Name(Name("DecoderEnrichments")), Importee.Name(Name("ObjectEncoderEnrichments")))))), Import(List(Importer(Term.Select(Term.Select(Term.Select(Term.Select(Term.Select(Term.Name("com"), Term.Name("twilio")), Term.Name("service")), Term.Name("encoding")), Term.Name("json")), Term.Name("SidFormats")), List(Importee.Wildcard())))), Import(List(Importer(Term.Select(Term.Select(Term.Select(Term.Name("_root_"), Term.Name("tests")), Term.Name("dtos")), Term.Name("Implicits")), List(Importee.Wildcard())))), Defn.Class(List(Mod.Case()), Type.Name("Something"), Nil, Ctor.Primary(Nil, Name(""), List(List(Term.Param(Nil, Term.Name("booleanValue"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("Boolean")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("stringValue"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("String")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("longProperty"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("Long")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("intProperty"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("Int")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("integerProperty"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("BigInt")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("floatProperty"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("Float")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("doubleProperty"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("Double")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("numberProperty"), Some(Type.Apply(Type.Name("Option"), List(Type.Name("BigDecimal")))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("property"), Some(Type.Apply(Type.Name("Option"), List(Type.Select(Term.Select(Term.Name("io"), Term.Name("circe")), Type.Name("Json"))))), Some(Term.Name("None"))), Term.Param(Nil, Term.Name("objectProperty"), Some(Type.Apply(Type.Name("Option"), List(Type.Select(Term.Select(Term.Name("io"), Term.Name("circe")), Type.Name("Json"))))), Some(Term.Name("None")))))), Template(Nil, Nil, Self(Name(""), None), Nil)), Defn.Object(Nil, Term.Name("Something"), Template(Nil, Nil, Self(Name(""), None), List(Defn.Val(List(Mod.Implicit()), List(Pat.Var(Term.Name("encodeSomething"))), None, Term.Block(List(Defn.Val(Nil, List(Pat.Var(Term.Name("readOnlyKeys"))), None, Term.Apply(Term.ApplyType(Term.Name("Set"), List(Type.Name("String"))), Nil)), Term.Apply(Term.Select(Term.Apply(Term.Apply(Term.Select(Term.Name("Encoder"), Term.Name("forProduct10")), List(Lit.String("boolean_value"), Lit.String("string_value"), Lit.String("long_property"), Lit.String("int_property"), Lit.String("integer_property"), Lit.String("float_property"), Lit.String("double_property"), Lit.String("number_property"), Lit.String("property"), Lit.String("object_property"))), List(Term.Function(List(Term.Param(Nil, Term.Name("o"), Some(Type.Name("Something")), None)), Term.Tuple(List(Term.Select(Term.Name("o"), Term.Name("booleanValue")), Term.Select(Term.Name("o"), Term.Name("stringValue")), Term.Select(Term.Name("o"), Term.Name("longProperty")), Term.Select(Term.Name("o"), Term.Name("intProperty")), Term.Select(Term.Name("o"), Term.Name("integerProperty")), Term.Select(Term.Name("o"), Term.Name("floatProperty")), Term.Select(Term.Name("o"), Term.Name("doubleProperty")), Term.Select(Term.Name("o"), Term.Name("numberProperty")), Term.Select(Term.Name("o"), Term.Name("property")), Term.Select(Term.Name("o"), Term.Name("objectProperty"))))))), Term.Name("mapJsonObject")), List(Term.Apply(Term.Select(Term.Placeholder(), Term.Name("filterKeys")), List(Term.Function(List(Term.Param(Nil, Term.Name("key"), None, None)), Term.ApplyUnary(Term.Name("!"), Term.ApplyInfix(Term.Name("readOnlyKeys"), Term.Name("contains"), Nil, List(Term.Name("key")))))))))))), Defn.Val(List(Mod.Implicit()), List(Pat.Var(Term.Name("decodeSomething"))), None, Term.Apply(Term.Apply(Term.Select(Term.Name("Decoder"), Term.Name("forProduct10")), List(Lit.String("boolean_value"), Lit.String("string_value"), Lit.String("long_property"), Lit.String("int_property"), Lit.String("integer_property"), Lit.String("float_property"), Lit.String("double_property"), Lit.String("number_property"), Lit.String("property"), Lit.String("object_property"))), List(Term.Eta(Term.Select(Term.Name("Something"), Term.Name("apply")))))))))))))
```

**Comment by olafurpg** [comment]
There be dragons. I see it currently assumes there are available `.tokens` so for synthetic trees that falls back to `TreeSyntax`. It should not call `.tokens` when the origin is empty (not parsed). Can you btw try scalameta v3.3.0? We experimented with changes in origins that resulted in a regression for `TreeSyntax` https://github.com/scalameta/scalameta/issues/1526

**Comment by blast-hardcheese** [comment]
I can do that now. Additionally, before we started this conversation, I started a branch to adjust the logic in `cantBeWrittenWithoutBackquotes` to match the scala spec. As it was very close by the time I noticed your response, I figured I should just finish it up, even if it doesn't get used:

https://github.com/scalameta/scalameta/compare/master...blast-hardcheese:1661-scalac-style-backquotes

**Comment by blast-hardcheese** [comment]
Same error with 3.3.0:

```
java.util.NoSuchElementException: None.get
        at scala.None$.get(Option.scala:349)
        at scala.None$.get(Option.scala:347)
        at org.scalafmt.internal.tokens.SyntaxTokensUtils$.$anonfun$commaSeparated0$1(SyntaxTokensUtils.scala:102)
        at scala.collection.Iterator$$anon$10.next(Iterator.scala:457)
        at scala.collection.Iterator.foreach(Iterator.scala:944)
        at scala.collection.Iterator.foreach$(Iterator.scala:944)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
        at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:59)
        at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:50)
        at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:186)
        at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:44)
        at scala.collection.TraversableOnce.to(TraversableOnce.scala:310)
        at scala.collection.TraversableOnce.to$(TraversableOnce.scala:308)
        at scala.collection.AbstractIterator.to(Iterator.scala:1432)
        at scala.collection.TraversableOnce.toList(TraversableOnce.scala:294)
        at scala.collection.TraversableOnce.toList$(TraversableOnce.scala:294)
        at scala.collection.AbstractIterator.toList(Iterator.scala:1432)
        at org.scalafmt.internal.tokens.SyntaxTokensUtils$.commaSeparated0(SyntaxTokensUtils.scala:101)
        at org.scalafmt.internal.tokens.SyntaxTokensUtils$.commaSeparated(SyntaxTokensUtils.scala:90)
        at org.scalafmt.internal.tokens.SyntaxTokensTerm$XtensionTermTupleSyntax$.tokensComma$extension(SyntaxTokensTerm.scala:224)
        at org.scalafmt.internal.tokens.SyntaxTokensTerm$XtensionTermTupleSyntax$.$u002C$extension(SyntaxTokensTerm.scala:235)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:229)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:226)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dApply$1(TreeDocOps.scala:67)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalafmt.internal.TreeDocOps.dApply(TreeDocOps.scala:67)
        at org.scalafmt.internal.TreeDocOps.dApply$(TreeDocOps.scala:66)
        at org.scalafmt.internal.TreePrinter.dApply(TreePrinter.scala:54)
        at org.scalafmt.internal.TreeDocOps.dApplyParen(TreeDocOps.scala:60)
        at org.scalafmt.internal.TreeDocOps.dApplyParen$(TreeDocOps.scala:59)
        at org.scalafmt.internal.TreePrinter.dApplyParen(TreePrinter.scala:54)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:286)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:182)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:278)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:287)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:54)
        at org.scalafmt.internal.TreeDocOps.dBlockI(TreeDocOps.scala:103)
        at org.scalafmt.internal.TreeDocOps.dBlockI$(TreeDocOps.scala:88)
        at org.scalafmt.internal.TreePrinter.dBlockI(TreePrinter.scala:54)
        at org.scalafmt.internal.TreeDocOps.dBlock(TreeDocOps.scala:86)
        at org.scalafmt.internal.TreeDocOps.dBlock$(TreeDocOps.scala:85)
        at org.scalafmt.internal.TreePrinter.dBlock(TreePrinter.scala:54)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:221)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dBodyO$2(TreeDocOps.scala:163)
        at scala.Option.fold(Option.scala:158)
        at org.scalafmt.internal.TreeDocOps.dBodyO(TreeDocOps.scala:161)
        at org.scalafmt.internal.TreeDocOps.dBodyO$(TreeDocOps.scala:160)
        at org.scalafmt.internal.TreePrinter.dBodyO(TreePrinter.scala:54)
        at org.scalafmt.internal.TreeDocOps.dBody(TreeDocOps.scala:158)
        at org.scalafmt.internal.TreeDocOps.dBody$(TreeDocOps.scala:157)
        at org.scalafmt.internal.TreePrinter.dBody(TreePrinter.scala:54)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:493)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:54)
        at org.scalafmt.internal.TreeDocOps.dBlockI(TreeDocOps.scala:103)
        at org.scalafmt.internal.TreeDocOps.dBlockI$(TreeDocOps.scala:88)
        at org.scalafmt.internal.TreePrinter.dBlockI(TreePrinter.scala:54)
        at org.scalafmt.internal.TreeDocOps.dBlock(TreeDocOps.scala:86)
        at org.scalafmt.internal.TreeDocOps.dBlock$(TreeDocOps.scala:85)
        at org.scalafmt.internal.TreePrinter.dBlock(TreePrinter.scala:54)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:444)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:523)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:287)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:54)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:374)
        at org.scalafmt.internal.SyntacticGroupOps$XtensionSyntacticGroup.wrap(SyntacticGroupOps.scala:15)
        at org.scalafmt.internal.TreeDocOps.$anonfun$dStats$1(TreeDocOps.scala:210)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalafmt.internal.TreeDocOps.dStats(TreeDocOps.scala:206)
        at org.scalafmt.internal.TreeDocOps.dStats$(TreeDocOps.scala:203)
        at org.scalafmt.internal.TreePrinter.dStats(TreePrinter.scala:54)
        at org.scalafmt.internal.TreePrinter.print(TreePrinter.scala:528)
        at org.scalafmt.internal.TreePrinter$.print(TreePrinter.scala:29)
        at com.twilio.guardrail.WriteTree$.$anonfun$unsafeWriteTree$1(WriteTree.scala:17)
```

**Comment by olafurpg** [comment]
If you have a TreeSyntax patch to fix this issue I propose we merge it and
open a separate issue to discuss the new printer.

**Comment by blast-hardcheese** [comment]
OK, I'll do some benchmarking to verify my new implementation doesn't slow things down too considerably, and we can talk about this tomorrow. Thanks @olafurpg!

**Comment by olafurpg** [comment]
You can open the pr and I can take a look tomorrow, performance is likely
fine.

**Comment by blast-hardcheese** [comment]
I also opened up `backtick-bugfix-backport-v3.7.x` which backports the change to 3.7.4 until 4.x is ready for prime-time

**Comment by olafurpg** [comment]
Thanks for fixing this @blast-hardcheese I discussed with @xeno-by and @MasseGuillaume and we agreed to move scala-syntax into the main scalameta repository and replace the current  tree printer. We opened a ticket and I've expanded on the details here https://github.com/scalameta/scalameta/issues/1680#issuecomment-402798970
